astropy/table/typing.py

```diff
@@ -0,0 +1,41 @@
+"""Typing for `astropy.table`."""
+
+from __future__ import annotations
+
+from typing import (
+    TYPE_CHECKING,
+    Generic,
+    Protocol,
+    Self,
+    SupportsIndex,
+    TypeVar,
+    overload,
+)
+
+if TYPE_CHECKING:
+    import numpy as np
+    from numpy.typing import NDArray
+
+    from astropy.utils.data_info import MixinInfo
+
+__all__ = [
+    "MixinColumn",
+]
+
+
+T = TypeVar("T")
+
+
+class MixinColumn(Protocol[T]):
+    info: MixinInfo
+    shape: tuple[int, ...]
+
+    @overload
+    def __getitem__(self, key: SupportsIndex) -> T: ...
+    @overload
+    def __getitem__(
+        self, key: slice | NDArray[np.bool] | NDArray[np.integer]
+    ) -> Self: ...
+    def __getitem__(self, key): ...
+    def __len__(self) -> int:
+        return self.shape[0]
```

Comment:
A docstring example, which will be checked by pytest-doctestplus, would serve as a good unit test.
I would suggest adding docstrings to the whole protocol, with doctests for unit tests.

Reply:
How do you mean ? this would also require adding default implementations for every method, right ? As far as I known, it's not something commonly done for `Protocol`s, so I'm hesitant to just go for it.

Reply:
e.g. 

```suggestion
        """Return the length of the mixin column.
        
        The MixinColumn provides a default implementation for the length based on the shape attribute.
        
        Examples
        ----------
        >>> from astropy.table.typing import MixinColumn
        
        >>> class Obj(MixinColumn):
        ...     shape = (5, 2, 3)
        
        >>> obj = Obj()
        >>> len(obj)
        5

        """"
        return self.shape[0]
```

Reply:
Oh I see, you're suggesting to demonstrate functionality through actual implementations ! While this indeed works as a test, I'm worried that it's also not a type-complete implementation and as such, should probably not be advertised to users through docstrings 🤔